run worker process in launcher pod

kuizhiqing opened the pull request:
**Feature**:
One can run worker process in the launcher pod by declare the same resources requirement for launcher and worker.

**Why this** ?
For a cluster fully equipped with GPU nodes, launcher pod occupied CPU resources which makes the worker cannot fully take advantage of all the physical resources.

With a cluster with CPU only nodes and GPU nodes, it may also suffer from weak efficiency due to the network between the two in most cases. 

**How** ?
Without introducing another configuration, we determine wether to run worker process in the launcher node by checking the resources declaration of launcher and worker. Worker in launcher is enabled if they are set to be equal.

There is definitely no need to set the same resources in practice if we do not run worker process in launcher.
> Actually, the launcher resources requirement should slightly more than that of worker theoretically, while we find it can be ignored in practice.

Don't forget to run `sshd` service in launcher in this mode.

This PR is a part of the project https://github.com/kubeflow/mpi-operator/issues/611.

alculquicondor commented:
If you really want to put a launcher process in the same node as a worker, you can simply set requests to zero, with the same behavior.

alculquicondor commented:
> Without introducing another configuration, we determine wether to run worker process in the launcher node by checking the resources declaration of launcher and worker. Worker in launcher is enabled if they are set to be equal.

This is not intuitive at all.

kuizhiqing commented:
> If you really want to put a launcher process in the same node as a worker, you can simply set requests to empty or something small, with the same behavior.

Well, there are some reason that worth notice,
1. In some strict managed platform, a pod without requests declaration may not be allowed to scheduler since it may encounter problem; in this case we can not leave it empty;
2. If we have a cluster only with GPU nodes, say one jobs takes 1000 nodes, if CPU launcher takes 2 cores, to make job homogeneous,  every nodes will reserve 2 cores each which is expansive;
3. If CPU launcher takes small resources, it is not efficient in the start phase while it waste of resource during the long training phase in most cases;
4. If we schedule the launcher to CPU only node, the network performance may not efficient since CPU-GPU node network usually bad than GPU-GPU node network.

I'm not sure if the issue https://github.com/kubeflow/mpi-operator/issues/503 require this feature, while it may related to this.

kuizhiqing commented:
> > Without introducing another configuration, we determine wether to run worker process in the launcher node by checking the resources declaration of launcher and worker. Worker in launcher is enabled if they are set to be equal.
> 
> This is not intuitive at all.

I'm agree that introducing a configure in CRD may be better, but the modification of CRD is somewhat expensive.

While personally, I totally OK with the statement that
> There is definitely no need to set the same resources in practice if we do not run worker process in launcher.

So I accept the design.

alculquicondor commented:
An explicit API will almost always be better. Maybe something like `useLauncherAsWorker: true`.

But ok, I can understand that your cluster might be entirely homogeneous, in which case a separate pod could be detrimental. 

Any concerns @tenzen-y ?

kuizhiqing commented:
@alculquicondor @tenzen-y 
Well, what's your opinion about this feature, should I continue to work with a version introducing `useLauncherAsWorker: true` or just hold it.

tenzen-y commented:
@alculquicondor @kuizhiqing First of all, I believe that this feature would be worth it. 
We can avoid falling into some rabbit holes due to mixed network protocols (well-known generic Ethernet and RoCE (or Infiniband)) since we could schedule both Launcher and Worker pods to the GPU node with a single network protocol (RoCE or Infiniband).

Actually, in my cluster, the launcher is scheduled to CPU nodes only with Ethernet, and the workers are scheduled to GPU nodes with Ethernet and RoCE networks. Provided that the mixed network protocols, I sometimes fall into rabbit holes :(

My only concern is the situation in which the launcher with the worker pod failed due to a worker process error. 
When the job uses horovod with elastic semantics, the point of failure would be increased, and then, the job is possible to lose the elasticity in the above situation. However, we can mention the disadvantage in the documentation to notice users.

> Well, what's your opinion about this feature, should I continue to work with a version introducing useLauncherAsWorker: true or just hold it.

I think we should introduce a new field for this feature in this PR.

As an alternative API name, we could add the `runLauncherWithinWorker: <bool>` and `launcherAsWorker: <bool>` in `.runPolicy`.

Or, I think we could create a new field `spawnStragtegy` under the `runPolicy` in the following:

```go
runPolicy:
    spawnStrategy:
        launcher: <asWorker (withinWorker)|Independent>
```

@alculquicondor @kuizhiqing WDYT?

alculquicondor commented:
I wouldn't touch `runPolicy` so that it stays the same as the other operators.

The name should match the behavior. What would you rather have?
1. The launcher is added to the hostfile, effectively making it an additional worker. We should allow to have zero explicit workers in this case.
2. There is no launcher job. The worker 0 has special logic in the entry-point.

I think option 1 is cleaner, because we don't have control over the entry-point. And we would have to re-implement retry semantics, as opposed to using the Job API.

tenzen-y commented:
> I wouldn't touch `runPolicy` so that it stays the same as the other operators.
> 
> The name should match the behavior. What would you rather have?
> 
> 1. The launcher is added to the hostfile, effectively making it an additional worker. We should allow to have zero explicit workers in this case.
> 2. There is no launcher job. The worker 0 has special logic in the entry-point.
> 
> I think option 1 is cleaner. Option 2 assumes that the operator has control over the entry-point, which is not true. And we would have to re-implement retry semantics, as opposed to using the Job API.

It sounds reasonable. I prefer option 1, too. 
@alculquicondor Does option 1 mean that MPIJob has nil `replicaSpec` for the `Worker`?  Or, the MPIJob has a `replicaSpec` with `replicas=0` for `Worker`?

alculquicondor commented:
> Does option 1 mean that MPIJob has nil replicaSpec for the Worker?

Yes, optionally it could be nil. Setting replicas=0 would require you to define a Pod spec, which is unnecessary.

kuizhiqing commented:
@tenzen-y 
I think we should not touch `runPolicy` since other operator or run mode do not have such config.

@alculquicondor 
> 1. The launcher is added to the hostfile, effectively making it an additional worker. We should allow to have zero explicit workers in this case.
While I prefer option 1.

OK, let me continue to work on it, tell me if you have more suggestions.

tenzen-y commented:
> > Does option 1 mean that MPIJob has nil replicaSpec for the Worker?
> 
> Yes, optionally it could be nil. Setting replicas=0 would require you to define a Pod spec, which is unnecessary.

SGTM

kuizhiqing commented:
@alculquicondor @tenzen-y  ready to review, PTAL

kuizhiqing commented:
@alculquicondor @tenzen-y  ready to review, PTAL

kuizhiqing commented:
/assign @alculquicondor 
/assign @tenzen-y

tenzen-y commented:
> /assign @alculquicondor /assign @tenzen-y

ACK
First of all, I can review this PR. Aldo will review this after my review.

tenzen-y commented:
@kuizhiqing Can you address CI issues?

kuizhiqing commented:
> @kuizhiqing Can you address CI issues?

Done

tenzen-y commented:
> > @kuizhiqing Can you address CI issues?
> 
> @tenzen-y Done

Sorry for the late response. I will come back here after the k/k enhancement freeze.
I have a lot of work...

tenzen-y commented:
I came back here right now.

tenzen-y commented on pkg/controller/mpi_job_controller.go:
```diff
@@ -628,15 +628,12 @@ func (c *MPIJobController) syncHandler(key string) error {
+		// The Intel and MPICH implementations require workers to communicate with the
+		// launcher through its hostname. For that, we create a Service which
+		// has the same name as the launcher's hostname.
+		if _, err = c.getOrCreateService(mpiJob, newLauncherService(mpiJob)); err != nil {
```
```suggestion
		if mpiJob.Spec.RunLauncherAsWorker || mpiJob.Spec.MPIImplementation == kubeflow.MPIImplementationIntel ||
			mpiJob.Spec.MPIImplementation == kubeflow.MPIImplementationMPICH {		
				if _, err = c.getOrCreateService(mpiJob, newLauncherService(mpiJob)); err != nil {
```

Why do we always need to create a Service for the Launcher?
IIUC, in the case of `runLauncherAsWorker=false`, creating service isn't needed.

kuizhiqing replied:
Done, I did want to keep it simple. It's OK this way.

tenzen-y commented on examples/v2beta1/pi/pi2.yaml:
I don't think that this example would be valuable. If we have this example, we need to have more examples like LauncherCreationPolicy.

So, instead of this example, could you add a document to the website repository?

kuizhiqing commented on pkg/controller/mpi_job_controller.go:
```diff
@@ -1291,6 +1302,13 @@ func updateDiscoverHostsInConfigMap(configMap *corev1.ConfigMap, mpiJob *kubeflo
 	var buffer bytes.Buffer
 	buffer.WriteString("#!/bin/sh\n")
+
+	// We donnot check if launcher is running here, launcher should always be there or the job failed
```
Since we leave the user start the launcher process and worker 0 sshd, it do NOT change the workflow.

kuizhiqing commented:
> Also, could you implement an integration test here: https://github.com/kubeflow/mpi-operator/blob/master/test/integration/mpi_job_controller_test.go?
> 
> Further more, could we support the zero explicit workers?
> 
> [#612 (comment)](https://github.com/kubeflow/mpi-operator/pull/612#issuecomment-1875682208)

@tenzen-y Thanks for your time on reviewing. I've addressed your comments, plz let me know if more modifs are needed.

We do support zero explicit workers setting, where I tested with the following example,
```
apiVersion: kubeflow.org/v2beta1
kind: MPIJob
metadata:
  name: pi
spec:
  runLauncherAsWorker: true
  slotsPerWorker: 1
  runPolicy:
    cleanPodPolicy: Running
    ttlSecondsAfterFinished: 60
  sshAuthMountPath: /home/mpiuser/.ssh
  mpiReplicaSpecs:
    Launcher:
      replicas: 1
      template:
        spec:
          containers:
          - image: mpioperator/mpi-pi:openmpi
            name: mpi-launcher
            securityContext:
              runAsUser: 1000
            command:
            - bash
            args:
            - -c
            - "/usr/sbin/sshd -f /home/mpiuser/.sshd_config && mpirun /home/mpiuser/pi"
            resources:
              limits:
                cpu: 1
                memory: 1Gi
```

And for the integration test part, after I finished it, I realize that this feature do NOT change the whole workflow as before, so it may take no different, while the unit test does.

Actually, this feature does the two things indeed,
* add svc for launcher
* add launcher to the hostfile

Note that
* the user should take care of the starting sshd service and run the real command in the launcher
* in the operator scope, workers do noting but start sshd service
then the introducing of this feature don't really change the RUNNING process but make it possible for the user by altering the configuration, say svc and hostfile.

tenzen-y commented on pkg/apis/kubeflow/v2beta1/types.go:
```diff
@@ -154,6 +154,12 @@ type MPIJobSpec struct {
 	// +kubebuilder:default:=1
 	SlotsPerWorker *int32 `json:"slotsPerWorker,omitempty"`
 
+	// RunLauncherAsWorker indicates wether to run worker process in launcher
```
```suggestion
	// RunLauncherAsWorker indicates whether to run worker process in launcher
```

tenzen-y commented on pkg/controller/mpi_job_controller.go:
```diff
@@ -656,13 +656,13 @@ func (c *MPIJobController) syncHandler(key string) error {
+		// The Intel and MPICH implementations require workers to communicate with the
+		// launcher through its hostname. For that, we create a Service which
+		// has the same name as the launcher's hostname.
+		if (mpiJob.Spec.RunLauncherAsWorker != nil && *mpiJob.Spec.RunLauncherAsWorker) ||
```
```suggestion
		if ptr.Deref(mpiJob.Spec.RunLauncherAsWorker, false) ||
```

tenzen-y commented on pkg/controller/mpi_job_controller_test.go:
```diff
@@ -151,7 +139,22 @@ func newMPIJobCommon(name string, startTime, completionTime *metav1.Time) *kubef
 func newMPIJob(name string, replicas *int32, startTime, completionTime *metav1.Time) *kubeflow.MPIJob {
 	mpiJob := newMPIJobCommon(name, startTime, completionTime)
-	mpiJob.Spec.MPIReplicaSpecs[kubeflow.MPIReplicaTypeWorker].Replicas = replicas
+	if *replicas > 0 {
```
```suggestion
	if ptr.Deref(replicas, 0) > 0 {
```

It should avoid potentially null pointer.

tenzen-y commented on pkg/controller/mpi_job_controller_test.go:
```diff
@@ -526,7 +529,8 @@ func TestAllResourcesCreated(t *testing.T) {
 				f.expectCreatePodAction(fmjc.newWorker(mpiJobCopy, i))
 			}
-			if implementation == kubeflow.MPIImplementationIntel ||
+			if (mpiJob.Spec.RunLauncherAsWorker != nil && *mpiJob.Spec.RunLauncherAsWorker) ||
```
```suggestion
			if ptr.Deref(mpiJob.Spec.RunLauncherAsWorker, false) ||
```

tenzen-y commented on pkg/controller/mpi_job_controller.go:
```diff
@@ -656,13 +656,13 @@ func (c *MPIJobController) syncHandler(key string) error {
+		// The Intel and MPICH implementations require workers to communicate with the
+		// launcher through its hostname. For that, we create a Service which
+		// has the same name as the launcher's hostname.
+		if (mpiJob.Spec.RunLauncherAsWorker != nil && *mpiJob.Spec.RunLauncherAsWorker) ||
+			mpiJob.Spec.MPIImplementation == kubeflow.MPIImplementationIntel ||
 			mpiJob.Spec.MPIImplementation == kubeflow.MPIImplementationMPICH {
```
Uhm, I'm wondering if we can make this condition another function like `workersCanHaveDedicatedService` since we're using the same condition in some places.

tenzen-y commented on pkg/controller/mpi_job_controller.go:
```diff
@@ -656,13 +656,13 @@ func (c *MPIJobController) syncHandler(key string) error {
 			}
 		}
-		if mpiJob.Spec.MPIImplementation == kubeflow.MPIImplementationIntel ||
+		// The Intel and MPICH implementations require workers to communicate with the
+		// launcher through its hostname. For that, we create a Service which
+		// has the same name as the launcher's hostname.
```
Please update this comment.

tenzen-y reviewed (commented):
@kuizhiqing Also, could you address my other comment below?

> Also, could you implement an integration test here: https://github.com/kubeflow/mpi-operator/blob/master/test/integration/mpi_job_controller_test.go?

tenzen-y commented:
> We do support zero explicit workers setting, where I tested with the following example,

Oh, I see. Thank you for the confirmation!

tenzen-y commented:
> And for the integration test part, after I finished it, I realize that this feature do NOT change the whole workflow as before, so it may take no different, while the unit test does.
> 
> Actually, this feature does the two things indeed,
> 
> add svc for launcher
> add launcher to the hostfile
> Note that
> 
> the user should take care of the starting sshd service and run the real command in the launcher
> in the operator scope, workers do noting but start sshd service
> then the introducing of this feature don't really change the RUNNING process but make it possible for the user by altering the configuration, say svc and hostile.

Uhm, it sounds reasonable considered in have a unit test, but let me know what @alculquicondor think.

tenzen-y commented:
@kuizhiqing Let me know once building errors are fixed.

kuizhiqing commented:
> @kuizhiqing Let me know once building errors are fixed.
> 
> 

@tenzen-y Done

tenzen-y reviewed (commented):
/lgtm
/assign @alculquicondor

kuizhiqing commented:
Thx @tenzen-y @alculquicondor , and I'll address the following in another PR
```
SA1019: "k8s.io/utils/pointer" is deprecated...
```

tenzen-y commented:
> Thx @tenzen-y @alculquicondor , and I'll address the following in another PR

lgtm

kuizhiqing commented on pkg/controller/mpi_job_controller.go:
```diff
@@ -656,13 +657,14 @@ func (c *MPIJobController) syncHandler(key string) error {
 			}
 		}
-		if mpiJob.Spec.MPIImplementation == kubeflow.MPIImplementationIntel ||
+		// If we want to run process in launcher, we should create a service for launcher.
```
The existing Service created with selector `training.kubeflow.org/job-role: worker` which can't be used directly. Without creating Service for launcher, we should change the service for worker and remove the selector.
Create a service is more preferable for me. WDYT @tenzen-y @alculquicondor

alculquicondor replied:
I think it's more efficient to just change the selector.

alculquicondor replied:
I would just change it for all

kuizhiqing replied:
@alculquicondor So it's a refactor of the service mechanism so that we create a job level service for all, no service for launcher and service for workers any more. Then we will use the job level service to access the target pod when necessary. Is that right ?

alculquicondor replied:
Yes. Although I'm not sure if IntelMPI would work with just one Service, as it's very picky about the hostname. Worth trying.

kuizhiqing replied:
OK,I will try it. While we just make the service refactor in this PR or with another one, since it's already a way to run with respect to the original service design?

alculquicondor replied:
here should be fine

tenzen-y replied:
SGTM

kuizhiqing replied:
I've add a service for both launcher and workers, for IntelMPI which need to access launcher with hostname, I modify the searches part in DNSConfig.

alculquicondor reviewed (commented):
Implementation looks good, but can you check something manually?

1. Create a long-running MPIJob using the released controller.
2. Upgrade the controller to a build containing this PR
3. Check that the MPIJob continues running.

I believe it should continue to run, but let's better be safe.

kuizhiqing commented:
> Implementation looks good, but can you check something manually?
> 
> 1. Create a long-running MPIJob using the released controller.
> 2. Upgrade the controller to a build containing this PR
> 3. Check that the MPIJob continues running.
> 
> I believe it should continue to run, but let's better be safe.

@alculquicondor 
I know that you want to make sure that a continuous upgrade should not break the existing job, so I did check it and it worked.
Note that one more svc with the same name as the job will be created for all the existing job, and it will not be used though. For new created job, only the new one svc will be created and be used.

alculquicondor commented:
ah, @tenzen-y already gave lgtm, so it will merge :)

tenzen-y commented:
> Implementation looks good, but can you check something manually?
> 
> 1. Create a long-running MPIJob using the released controller.
> 2. Upgrade the controller to a build containing this PR
> 3. Check that the MPIJob continues running.
> 
> I believe it should continue to run, but let's better be safe.

Maybe we can add an integration test for this situation by enabling and disabling this feature, but I'm ok with working on it at another PR.

tenzen-y reviewed (commented):
/lgtm
